chainer/functions/loss/discriminative_loss.py

```diff
@@ -0,0 +1,248 @@
+import cupy as cp
+import numpy as np
+
+import chainer.functions as F
+
+
+class DiscriminativeMarginBasedClusteringLoss:
+    """ Discriminative margin based clustering loss function
+
+    This is the implementation of following paper: 
+    https://arxiv.org/pdf/1802.05591.pdf
+
+    In segmentation, one of the biggest problem is having noise at the output 
+    of trained network.
+    For cross-entropy based approaches, if the pixel value is wrong,
+    the loss value will be same independent from wrong pixel's location.
+    However, for segmentation, even though network gives wrong pixel output, 
+    it is desirable to have it
+    as close as possible to the original position.
+    By applying discriminative loss function, groups of segmentation instances 
+    can be moved together.
+
+    This loss function calculates three different parameters:
+        - Variance Loss:
+            Loss to penalize distances between pixels which are belonging 
+            to same instance. (Pull force)
+        - Distance loss:
+            Loss to penalize distances between the centers of instances. 
+            (Push force)
+        - Regularizer loss:
+            Small regularization loss to penalize weights against overfit
+
+    :param delta_v: Used to define minimum distance to start penalizing variance
+    :param delta_d: Used to define maximum distance to stop penalizing distance
+    :param max_n_clusters: Used to define maximum possible number of clusters.
+    :param norm: Norm to calculate distance between pixels and cluster centers
+    :param alpha: Weight for variance loss      (alpha * variance_loss)
+    :param beta: Weight for distance loss       (beta * distance_loss)
+    :param gamma: Weight for regularizer loss   (gamma * regularizer_loss)
+
+    :return 
+    loss: 
+        (alpha * variance_loss) + 
+        (beta * distance_loss) + 
+        (gamma * regularizer_loss)
+
+    """
+
+    def __init__(self, delta_v, delta_d, max_n_clusters, norm=1, alpha=1.0,
+                 beta=1.0, gamma=0.001):
+        self.delta_v = delta_v
+        self.delta_d = delta_d
+        self.alpha = alpha
+        self.beta = beta
+        self.gamma = gamma
+        self.max_n_clusters = max_n_clusters
+
+        # L1 or L2 norm is allowed only
+        if norm != 1:
+            raise Exception(
+                'Currently discriminative loss only supports L1 Norm')
+
+        if norm == 2:
```

Comment:
I think this condition is never held because `norm` should be `1` by the previous `if` statement. Do you want to do some thing like this?

```python
if norm == 1:
    self.norm = self.l1_norm
elif norm == 2:
    self.norm = self.l2_norm
else:
    raise Exception(...)
```